crates/ruma-events/src/room/message/content_serde.rs

```diff
@@ -76,6 +84,67 @@ impl<'de> Deserialize<'de> for MessageType {
     }
 }
 
+/// Helper struct to determine the itemtype from a `serde_json::value::RawValue`
+#[derive(Debug, Deserialize)]
+#[cfg(feature = "unstable-msc4274")]
+struct ItemTypeDeHelper {
+    /// The item type field
+    itemtype: String,
+}
+
+#[cfg(feature = "unstable-msc4274")]
+impl<'de> Deserialize<'de> for GalleryItemType {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let json = Box::<RawJsonValue>::deserialize(deserializer)?;
+        let ItemTypeDeHelper { itemtype } = from_raw_json_value(&json)?;
+
+        Ok(match itemtype.as_ref() {
+            "m.audio" => Self::Audio(from_raw_json_value(&json)?),
+            "m.file" => Self::File(from_raw_json_value(&json)?),
+            "m.image" => Self::Image(from_raw_json_value(&json)?),
+            "m.video" => Self::Video(from_raw_json_value(&json)?),
```

Comment:
I find it weird that those don't complain about a missing `msgtype`, but that's good for us.